### DISCUSSION TITLE
Style editing performance tips

### DISCUSSION THREAD

**Discussion by orangemug** [openlayers/ol-mapbox-style#1052] [General]
When building an app with `maplibre-gl-js`/`mapbox-gl-js` you can just throw a new style at it and it does an "internal diff" updating the style from what has changed. This appears to be slow in `ol-mapbox-style` currently I'm just calling

```js
apply(mapInstance, newStyle)
```

In `maplibre-gl-js` only the layer that's updated changes. Whereas we get a full, fairly slow, re-render in `ol-mapbox-style`. Anybody any ideas would I go about making this quicker?

**Comment by ahocevar** [comment]
Implementing a complete internal diff mechanism is planned, but lacks funding or code contributions. But some important building blocks of an internal diff are already available and can help with the use cases you mentioned:

[`updateMapboxLayer(mapInstance, newLayerStyle)`](https://github.com/openlayers/ol-mapbox-style#updatemapboxlayer)
[`addMapboxLayer(mapInstance, newLayerStyle, beforeLayerId)`](https://github.com/openlayers/ol-mapbox-style#addmapboxlayer)
[`removeMapboxLayer(mapInstance, layerStyleOrId)`](https://github.com/openlayers/ol-mapbox-style#removemapboxlayer)
[`updateMapboxSource(mapInstance, sourceId, newSourceDef)`](https://github.com/openlayers/ol-mapbox-style#updatemapboxsource)

**Reply by orangemug** [reply]
Maybe a first step could be instead of a total style diff, either

 - Object equality - apps often use immutable objects anyway so `layer === newLayer` would do
 - `shouldLayerUpdate` hook - to defer updates, allow the controlling app to decide to defer render

Either of those could be a first step towards what (maplibre/mapbox)-gl-js does.

**Reply by ahocevar** [reply]
There is a diff module in the style-spec package, and the aforementioned functions are already there. The task ahead is combining the two to achieve a diff. If you can contribute code or funds, it would be appreciated to move this effort forward.

**Reply by orangemug** [reply]
Thanks, I don't have any funds at the moment as I'm between jobs hacking on Maparatus. I'd like to contribute to this as a part of that work, but other tasks in the short term will take priority, like https://github.com/openlayers/ol-mapbox-style/issues/10#issuecomment-1850471621.
